src/coord/coord.rs

```diff
@@ -1650,8 +1650,63 @@ where
         self.import_source_or_view(id, &index.on, &mut dataflow);
         dataflow.add_index_to_build(*id, index.on.clone(), on_type.clone(), index.keys.clone());
         dataflow.add_index_export(*id, index.on, on_type, index.keys.clone());
-        self.broadcast_dataflow_creation(dataflow);
         self.insert_index(*id, &index, self.logical_compaction_window_ms);
+        self.validate_dataflow(&mut dataflow);
+        self.broadcast_dataflow_creation(dataflow);
+    }
+
+    /// Performs some sanity checking on the dataflow before shipping it.
+    ///
+    /// In particular, there are requirement on the `as_of` field for the dataflow
+    /// and the `since` frontiers of created arrangements, as a function of the `since`
+    /// frontiers of dataflow inputs (sources and imported arrangements).
+    fn validate_dataflow(&mut self, dataflow: &mut DataflowDesc) {
+        use differential_dataflow::lattice::Lattice;
+        use timely::progress::timestamp::Timestamp;
+
+        // The identity for `join` is the minimum element.
+        let mut since = Antichain::from_elem(Timestamp::minimum());
+
+        // TODO: Populate "valid from" information for each source.
+        // For each source, ... do nothing because we don't track `since` for sources.
+        // for (instance_id, _description) in dataflow.source_imports.iter() {
+        //     // TODO: Extract `since` information about each source and apply here. E.g.
+        //     since.join_assign(&self.source_info[instance_id].since);
+        // }
+
+        // For each imported arrangement, lower bound `since` by its own frontier.
+        for (global_id, (_description, _typ)) in dataflow.index_imports.iter() {
+            since.join_assign(
+                self.indexes
+                    .since_of(global_id)
+                    .expect("global id missing at coordinator"),
+            );
+        }
+
+        // For each produced arrangement, force its compaction frontier to be at least `since`.
+        for (global_id, _description, _typ) in dataflow.index_exports.iter() {
+            self.indexes
+                .get_mut(global_id)
+                .expect("global id missing at coordinator")
+                .advance_since(&since);
+        }
+
+        // TODO: Produce "valid from" information for each sink.
+        // For each sink, ... do nothing because we don't yield `since` for sinks.
+        // for (global_id, _description) in dataflow.sink_exports.iter() {
+        //     // TODO: assign `since` to a "valid from" element of the sink. E.g.
+        //     self.sink_info[global_id].valid_from(&since);
+        // }
+
+        // Ensure that the dataflow's `as_of` is at least `since`.
+        if let Some(as_of) = &dataflow.as_of {
+            // If we have requested a specific time that is invalid .. someone errored.
+            use timely::order::PartialOrder;
+            assert!(<_ as PartialOrder>::less_equal(&since, &as_of));
```

Comment:
This logic makes sense / I had not considered this but maybe we should error out / emit a log and not assert here? Not sure how to balance being loud when we break invariants vs having fires

Reply:
It's a legit question. At this point (validation) this means that we have actively requested a specific `as_of` and failed to validate it at that time. I think it's fair for this to be a panic, as it 100% indicates that we've made a mistake.

I'm tracking down such a mistake at the moment, and can report back what I learn about that process.

Reply:
👍

Reply:
Ah yes. So our sinks have potentially been reporting partially compacted data out, for example. They explicitly request an `as_of` frontier of `[0]`, but should probably reflect on the `since` bounds of their inputs.